src/scportrait/pipeline/extraction.py

```diff
@@ -524,206 +509,145 @@ def _extract_classes(self, px_center, arg, return_failed_ids=False):
+                    axs[i].axis("off")
+                fig.tight_layout()
+                fig.show()
 
-            if return_failed_ids:
-                return [save_index]
-            else:
-                return None
+        if return_results:
+            return save_index, stack, cell_id
+        else:
+            self._single_cell_data[save_index] = stack
+            self._single_cell_index[save_index] = [save_index, cell_id]
+            return None
 
-    def _extract_classes_multi(self, px_centers, arg_list):
+    def _extract_classes_multi(
+        self,
+        arg_list,
+    ):
         self._setup_normalization()
 
         self.seg_masks = mmap_array_from_path(self.path_seg_masks)
         self.image_data = mmap_array_from_path(self.path_image_data)
 
-        self._tmp_single_cell_index = mmap_array_from_path(self._tmp_single_cell_index_path)
-        self._tmp_single_cell_data = mmap_array_from_path(self._tmp_single_cell_data_path)
-
-        results = []
-        for arg in arg_list:
-            x = self._extract_classes(px_centers, arg, return_failed_ids=True)
-            results.append(x)
-
-        return flatten(results)
-
-    def _transfer_tempmmap_to_hdf5(self):
-        self.log("Transferring results to final HDF5 data container.")
-
-        # reconnect to memory mapped temp arrays
-        _tmp_single_cell_index = mmap_array_from_path(self._tmp_single_cell_index_path)
-        _tmp_single_cell_data = mmap_array_from_path(self._tmp_single_cell_data_path)
-
-        self.log(f"number of cells too close to image edges to extract: {len(self.save_index_to_remove)}")
+        # get processing results
+        results = [self._extract_classes(arg, return_results=True) for arg in arg_list]
 
-        # generate final index of all of the rows that we wish to keep out of the original array
-        keep_index = np.setdiff1d(np.arange(_tmp_single_cell_index.shape[0]), self.save_index_to_remove)
+        return results
 
-        # get cell_ids of the cells that were successfully extracted
-        _, cell_ids = _tmp_single_cell_index[keep_index].T
-        _, cell_ids_removed = _tmp_single_cell_index[self.save_index_to_remove].T
+    def _write_to_hdf5(self, results, hdf5_lock):
+        """Function for writing results to HDF5 file in a thread-safe manner."""
+        with hdf5_lock:
+            with h5py.File(self.output_path, "a") as hf:
+                self._single_cell_data = hf["single_cell_data"]
+                self._single_cell_index = hf["single_cell_index"]
 
-        # convert to correct type
-        cell_ids = cell_ids.astype(self.DEFAULT_SEGMENTATION_DTYPE)
-        cell_ids_removed = cell_ids_removed.astype(self.DEFAULT_SEGMENTATION_DTYPE)
+                for res in results:
+                    save_index, stack, cell_id = res
+                    self._single_cell_data[save_index] = stack
+                    self._single_cell_index[save_index] = [save_index, cell_id]
```

Comment:
Two things here:

1. Now that I understand there are more users than just people looking to accelerate things locally, I think that making this generic over zarr/hdf5 storage types would be great.  People could then use this from the cloud.  You can definitely add a benchmark to the repo and explain advantages/disadvantages.
2. I think using types everywhere would really enforce/make easier doing 1.

Reply:
1. absolutely agree, that would definitely improve the package.
2. When you mean using types everywhere what do you mean exactly?

From a coding point of view (also for ease of maintenance going forward) what would be the fundamental changes you would suggest to implement in our package to achieve this goal? I guess this is the part where I as a software developing newbie(ish) am sometimes a bit lost about the best way to approach these types of problems.

Reply:
So I added doc strings and type hints to everything in the extraction.py file. It's been on my radar for a bit but I haven't gotten around to doing it retroactively for all code in the repo. Does this address your point 2 @ilan-gold?

Reply:
I've made a separate git issue as a placeholder to track discussions about 1. This would become too large for this PR. see #217